### DISCUSSION TITLE
SpanPoint API to use ESP-now with other sources and auto-pairing

### DISCUSSION THREAD

**Discussion by fhorlaville** [HomeSpan/HomeSpan#472] [Q&A]
Hi - a big THANK YOU and CONGRATULATIONS on the HomeSpan work, it it truly fantastic.

I have two questions on the SpanPoint implementation:

1. could you share an API or an implementation guide so that non-HomeSpan sensors could communicate with a SpanPoint instance running on the master? I'm thinking specifically about ESP8266 remotes which cannot run HomeSpan (e.g Wemos D1 or ESP-01).
2. will you be implementing some form of [auto-pairing](https://randomnerdtutorials.com/esp-now-auto-pairing-esp32-esp8266/) ?

Many thanks, happy new year

**Comment by HomeSpan** [comment]
Thanks -- I'm really glad you are finding HomeSpan useful.

Re the API, it's just straight ESP-NOW but with encryption implemented.  I'll produce a short write-up of the specifics and post in the next few days.

I don't have any ESP8266 chips, but since questions about this come up from time to time, I'll order one and see if I can port SpanPoint as a standalone module.

I've thought about auto-pairing and there some ways of doing this, but I'm not sure how useful this would really be in a non-commercial setting (i.e. not mass-producing the boards) since you still need to create sketches for each specific Accessory.

**Reply by fhorlaville** [reply]
(Happy Ito help on the port)

There are intermediate situations where even in a non-commercial setting you may not have a full IDE access. I've created a lamp for my daughter as a Christmas gift - she will be using the button / LED system to connect to her own WIFI without needing to actually reprogram the chip. Having the same kind of ease of use for "satellite" pairing would allow me to create weather stations for all members of the family and throw in as many additional room sensors as they want without having to keep track of each individual MAC address. Typical use case would follow your "dynamic lights" example where one could add as many sensor modules to the main HomeSpan without ever needing to re-flash the firmware.

**Reply by flohse123** [reply]
A SpanPoint solution for ESP8266 as remote sensor would be great since I have quite a lot of them in my system and would love to change them from http to ESP-NOW. 

In the meantime: Could I just implement the usual ESP-NOW commands in the code without using the beautiful simplicity of HomeSpan? Are the needed libraries already implemented or would I have to include them?

**Reply by HomeSpan** [reply]
I have just added new functionality that allows HomeSpan to receive ESP-NOW data from an ESP8266.  The new functionality (which consists of a 5th optional parameter in the SpanPoint constructor) is in the dev branch --- see the dev branch version of the SpanPoint page for details.  I also included on that page a detailed example sketch, *RemoteDevice8266*, showing how to implement ESP-NOW on an ESP8266 so it can properly communicate with an ESP32 running HomeSpan.

**Reply by flohse123** [reply]
Thank you indeed for the effort. I just tried to implement this but cannot figure out the correct size for the received data. If I set it too high, I get the following error:

> SpanPoint Warning! 44 bytes received from 3C:61:05:D0:DA:4B does not match 45-byte queue size

If I set it one byte lower (or use the sizeof()-function of the struct I created) it throws a Guru Meditation error and reboots. Here ist the error message:

> Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.
> 
> Core  1 register dump:
> PC      : 0x4008b358  PS      : 0x00060130  A0      : 0x8016f480  A1      : 0x3ffb21e0  
> A2      : 0x00000005  A3      : 0x00000001  A4      : 0x000000ff  A5      : 0x0000ff00  
> A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0x2a2a0a00  A9      : 0x00000000  
> A10     : 0xffffff80  A11     : 0x00060523  A12     : 0x00060520  A13     : 0x00060523  
> A14     : 0x007e5e84  A15     : 0x003fffff  SAR     : 0x0000001d  EXCCAUSE: 0x0000001c  
> EXCVADDR: 0x00000005  LBEG    : 0x4008b379  LEND    : 0x4008b389  LCOUNT  : 0xffffffff  
> 
> Backtrace: 0x4008b355:0x3ffb21e0 0x4016f47d:0x3ffb21f0 0x4017650a:0x3ffb2500 0x40176546:0x3ffb2590 0x400ebc16:0x3ffb25d0 0x400d749d:0x3ffb2670 0x400e0111:0x3ffb2700 0x400e02ba:0x3ffb27e0 0x400d7900:0x3ffb2800 0x400ed019:0x3ffb2820

Any ideas what might be the problem?

**Reply by HomeSpan** [reply]
Check to make sure that the data structure that receives the data is of sufficient size to store 44 bytes.  When there is a mismatch in the size of the received data and the size expected by SpanPoint, the data is never copied by the get() function. By lowering the number to correctly match 44 bytes, my guess is that the get() function is now able to copy the data into your structure, but is finding there is insufficient space.

**Reply by flohse123** [reply]
Thank you. The error was a LOG1-Statement with %s instead of %d for an integer. I don't understand why compilers do not check this kind of stuff. But now it's working like a charm. Thank you so much for this incredible library.
